virtcontainers/container.go

```diff
@@ -416,63 +416,78 @@ func newContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
 	return c, nil
 }
 
+// rollbackFailingContainerCreation rolls back important steps that might have
+// been performed before the container creation failed.
+// - Unplug CPU and memory resources from the VM.
+// - Unplug devices from the VM.
+func rollbackFailingContainerCreation(c *Container, err error) {
+	if err != nil && c != nil {
+		if err2 := c.removeResources(); err2 != nil {
+			c.Logger().WithError(err2).Error("rollback failed removeResources()")
+		}
+		if err2 := c.detachDevices(); err2 != nil {
+			c.Logger().WithError(err2).Error("rollback failed detachDevices()")
+		}
+		if err2 := c.removeDrive(); err2 != nil {
+			c.Logger().WithError(err2).Error("rollback failed removeDrive()")
+		}
+	}
+}
+
 // createContainer creates and start a container inside a Pod. It has to be
 // called only when a new container, not known by the pod, has to be created.
-func createContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
+func createContainer(pod *Pod, contConfig ContainerConfig) (c *Container, err error) {
 	if pod == nil {
 		return nil, errNeedPod
 	}
 
-	c, err := newContainer(pod, contConfig)
+	c, err = newContainer(pod, contConfig)
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	if err := c.createContainersDirs(); err != nil {
-		return nil, err
+	if err = c.createContainersDirs(); err != nil {
+		return
 	}
 
-	if !c.pod.config.HypervisorConfig.DisableBlockDeviceUse {
-		agentCaps := c.pod.agent.capabilities()
-		hypervisorCaps := c.pod.hypervisor.capabilities()
+	// In case the container creation fails, the following takes care
+	// of rolling back all the actions previously performed.
+	defer rollbackFailingContainerCreation(c, err)
 
-		if agentCaps.isBlockDeviceSupported() && hypervisorCaps.isBlockDeviceHotplugSupported() {
```

Comment:
@sboeuf I know this has been merged, but I hadnt taken a look at this PR. Any reason why the checks for block device support were taken out here?